### PR TITLE
[IMP] spreadhseet: add geo charts

### PR DESCRIPTION
**Pull request by hokolomopo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/spreadsheet/static/topojson/world_country_iso_mapping.json
[the remark this answers is not in the document]
Comment (LucasLefevre) [reply addons/spreadsheet/static/topojson/world_country_iso_mapping.json:1]: > It depends wether we allos to create geoChart on spreadsheet cells, or only for odoo data I guess

Data from odoo is the main use case we want to support

File: addons/spreadsheet/static/src/public_readonly_app/public_readonly.js [hunk 1]
```diff
@@ -61,7 +63,10 @@ export class PublicReadonlySpreadsheet extends Component {
         this.data = await this.http.get(this.props.dataUrl);
         this.model = new Model(
             this.data,
-            { mode: this.props.mode === "dashboard" ? "dashboard" : "readonly" },
+            {
+                mode: this.props.mode === "dashboard" ? "dashboard" : "readonly",
+                external: { geoJsonService: this.geoJsonService },
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply addons/spreadsheet/static/src/public_readonly_app/public_readonly.js:68]: I made it "work" for public user (chart is converted to an image). But I forgot to remove this :D

File: addons/spreadsheet/static/src/helpers/geo_json_service.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        return featureMappingCache.get(region);
+    }
+
+    if (region === "usa") {
+        if (usStatesMappingPromise) {
+            return usStatesMappingPromise;
+        }
+        usStatesMappingPromise = orm
+            .searchRead("res.country.state", [["country_id.code", "=", "US"]], ["name", "code"])
+            .then((usStates) => {
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply addons/spreadsheet/static/src/helpers/geo_json_service.js]: Idk, neither are great syntax IMHO. 

image

[the remark this answers is not in the document]
Comment (LucasLefevre) [reply addons/spreadsheet/static/src/helpers/geo_json_service.js]: Oh yeah, indeed, if you store the promise in a variable it's not better

File: addons/spreadsheet/static/src/assets_backend/helpers.js [hunk 3]
```diff
@@ -1,9 +1,10 @@
-import { loadBundle } from "@web/core/assets";
+import { loadBundle, loadJS } from "@web/core/assets";
 
 /**
  * Load external libraries required for o-spreadsheet
  * @returns {Promise<void>}
  */
 export async function loadSpreadsheetDependencies() {
     await loadBundle("web.chartjs_lib");
+    await loadJS("/spreadsheet/static/lib/chartjs-chart-geo/chartjs-chart-geo.js");
```
Comment (LucasLefevre) [inline addons/spreadsheet/static/src/assets_backend/helpers.js:9]: that's funny.
We could just `loadBundle("spreadsheet.dependencies")`, right ?
Or even include `spreadsheet.dependencies` in `spreadsheet.o_spreadsheet` bundle directly if we are fetching those libs in all cases

File: addons/spreadsheet/static/src/helpers/geo_json_service.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                });
+
+            geoJsonPromises.set(url, promise);
+            return promise;
+        }
+
+        const stateNameRegex = /(.*?)(\(.*\))?$/;
+
+        return {
+            getAvailableRegions: () => [
```
Comment (LucasLefevre) [inline addons/spreadsheet/static/src/helpers/geo_json_service.js:117]: our friends from south america are gonna be sad

Comment (hokolomopo) [reply addons/spreadsheet/static/src/helpers/geo_json_service.js:117]: You're right, let's make this PR 50K lines 😎

**Review by aab-odoo — COMMENTED** [review]
you added the lib in both spreadsheet and web, afaics (I first saw it in web as I've been notified, and my first reaction was "why isn't it in spreadsheet").

**Comment by hokolomopo** [comment]
> spreadsheet

Yeah sorry I just saw it as well 😅 But imagine geo chart in odoo views .... 👀

**Comment by pro-odoo** [comment]
https://github.com/odoo/odoo/pull/191109

**Pull request closed without merging**